File: 3_Arrays/Sortings/quick sort/quick sort.c
#include <malloc.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void swap(int *array, int a, int b) {
    int temp = array[a];
    array[a] = array[b];
    array[b] = temp;
}

int partition(int *array, int begin, int end) {
    int pivot = array[end];
    int i = begin - 1;
    for (int j = begin; j < end; j++)  
        if (array[j] <= pivot) 
            swap(array, j, ++i);
    
    swap(array, end, ++i);
    return i;
}

void quick_sort(int *array, int begin, int end) {
    if (begin < end) {
        int i = partition(array, begin, end);
        quick_sort(array, begin, i-1);
        quick_sort(array, i+1, end);
    }
}

int main() {
    int len = 20;
    int *array = (int*)calloc(sizeof(int), len);

    // fill with random elements
    srand(time(NULL));
    for (int i = 0; i < len; i++) 
        array[i] = rand() % 30;

    // print
    for (int i = 0; i < len; i++) 
        printf("%d ", array[i]);
    printf("\n");

    quick_sort(array, 0, len-1);

    // print
    for (int i = 0; i < len; i++) 
        printf("%d ", array[i]);
    printf("\n");

    return 0;
}

File: 3_Arrays/Sortings/quick sort/quick sort.js
function swap(array, a, b) {
    let temp = array[a];
    array[a] = array[b];
    array[b] = temp;
}

function partition(array, begin, end) {
    let pivot = array[end];
    let i = begin - 1;
    for (let j = begin; j < end; j++)  
        if (array[j] <= pivot) 
            swap(array, j, ++i);
    
    swap(array, end, ++i);
    return i;
}

function quickSort(array, begin, end) {
    if (begin < end) {
        let i = partition(array, begin, end);
        quickSort(array, begin, i-1);
        quickSort(array, i+1, end);
    }
}


let arr = new Array(4)
for (let i = 0; i < arr.length; i++) 
    arr[i] = Math.floor(Math.random() * 10) 

console.log(arr)
quickSort(arr, 0, arr.length-1);
console.log(arr)
